bazel_tools/scala.bzl

```diff
@@ -201,6 +201,121 @@ def _create_scala_source_jar(**kwargs):
             srcs = kwargs["srcs"],
         )
 
+def _scaladoc_jar_impl(ctx):
+    srcFiles = [
+        src.path
+        for src in ctx.files.srcs
+    ]
+    pluginPaths = []
+    for p in ctx.attr.plugins:
+        if hasattr(p, "path"):
+            pluginPaths.append(p)
+        elif hasattr(p, "scala"):
+            pluginPaths.extend([j.class_jar for j in p.scala.outputs.jars])
+        elif hasattr(p, "java"):
+            pluginPaths.extend([j.class_jar for j in p.java.outputs.jars])
+            # support http_file pointed at a jar. http_jar uses ijar,
+            # which breaks scala macros
+
+        elif hasattr(p, "files"):
+            pluginPaths.extend([f for f in p.files if "-sources.jar" not in f.basename])
+
+    transitive_deps = [dep[JavaInfo].transitive_deps for dep in ctx.attr.deps]
+    classpath = depset([], transitive = transitive_deps).to_list()
+
+    outdir = ctx.actions.declare_directory(ctx.label.name + "_scaladoc_tmpdir")
+
+    ctx.actions.run(
+        executable = ctx.executable._scaladoc,
+        inputs = ctx.files.srcs + classpath + pluginPaths,
+        outputs = [outdir],
+        arguments = [
+            "-d",
+            outdir.path,
+            "-classpath",
```

Comment:
You can avoid building potentially large strings in Starlark by using [`Args` objects](https://docs.bazel.build/versions/master/skylark/lib/Args.html). E.g.
```
args = ctx.actions.args()
args.add_all(["-d", outdir.path])
args.add("-classpath")
args.add_joined(classpath, join_with=":")
args.add_joined(pluginPaths, join_with=",", format_joined="-Xplugins:%s")
...

ctx.actions.run(
    ...
    arguments = [args],
    ...
```

Reply:
Good idea.